floodscan_validation/exploration/threshold_deciding.R

```diff
@@ -0,0 +1,1083 @@
+  geom_bar(
+    stat = "identity"
+  ) +
+  facet_wrap(
+    ~ HYBAS_LEVEL,
+    nrow = 1,
+    scales = "free",
+    labeller = labeller(
+      HYBAS_LEVEL = \(x) paste0("Basin level: ", x)
+    )
+  ) +
+  labs(
+    x = "Basin",
+    y = "# of gauges",
+    title = "Gauages per basin, based on basin level"
+  )
+
+gauges_map <- ggplot(
+  sf_basins
+) +
+  geom_sf() +
+  geom_sf(
+    data = sf_gauges
+  ) +
+  facet_wrap(
+    ~ HYBAS_LEVEL,
+    labeller = labeller(
+      HYBAS_LEVEL = \(x) paste0("Basin level: ", x)
+    )
+  ) +
+  coord_sf(
+    datum = NA
+  ) +
+  labs(
+    title = "Map of hydrobasins and gauges"
+  )
+
+gauges_map +
+  basin_bar_chart +
+  plot_layout(
+    ncol = 1
+  )
+
+########################################
+#### GET FLOODSCAN THRESHOLD LEVELS ####
+########################################
+
+# simple approach, take max avlue of each year we have full data for
+# and then take the median, which will ensure half the years would trigger
+# and half would not for an approx 2 year RP thresholding
+fs_gauge_thresholds <- fs_gauge %>%
+  mutate(
+    year = lubridate::year(time)
+  ) %>%
+  filter(
+    year %in% 1999:2022
+  ) %>%
+  group_by(
+    HYBAS_ID,
+    year
+  ) %>%
+  summarize(
+    sfed_max = max(sfed),
+    .groups = "drop_last"
+  ) %>%
+  summarize(
+    sfed_threshold = median(sfed_max),
+    .groups = "drop"
+  )
+
+# now get the times that these gauges have crossed the threshold
+
+fs_gauge_activations <- fs_gauge %>%
+  left_join(
+    fs_gauge_thresholds,
+    by = "HYBAS_ID"
+  ) %>%
+  mutate(
+    fs_above_threshold = sfed >= sfed_threshold
+  )
+
+# now find when the google data has passed their thresholds
+
+google_gauge_activations <- google_reanalysis %>%
+  mutate(
+    gauge_id = paste0("hybas_", hybas_station)
+  ) %>%
+  left_join(
+    google_rp,
+    by = "gauge_id"
+  ) %>%
+  mutate(
+    google_activation_2y_rp = prediction >= `2_year_return_period`,
+    google_activation_5y_rp = prediction >= `5_year_return_period`,
+    google_activation_20y_rp = prediction >= `20_year_return_period`,
+  )
+
+# now bring together for validation
+# we will consider an activation valid if within 7 days of the google gauge
+# reaching the 2 year RP, the floodscan around the gauge also did
+
+gauge_validation <- google_gauge_activations %>%
+  select(
+    hybas_station,
+    time,
+    google_activation_2y_rp
+  ) %>%
+  inner_join(
+    select(
+      fs_gauge_activations,
+      hybas_station = HYBAS_ID,
+      time,
+      fs_above_threshold
+    ),
+    by = c("hybas_station", "time")
+  ) %>%
+  group_by(
+    hybas_station
+  ) %>%
+  mutate(
+    fs_above_threshold_7_days = zoo::rollsum(
+      x = fs_above_threshold,
+      k = 7,
+      fill = FALSE,
+      align = "left"
+    ) > 0,
+    google_id = cumsum(google_activation_2y_rp != lag(google_activation_2y_rp, default = FALSE)),
+    fs_id = cumsum(fs_above_threshold_7_days != lag(fs_above_threshold_7_days, default = FALSE)),
```

Comment:
Yes, good idea, will do the sorting just as an in case! I see what you mean on the inclusion of consecutive rises/falls, will see how that applies to the basins.